src/main/daml/Daml/Finance/Settlement/Batch.daml

```diff
@@ -52,25 +54,35 @@ template Batch
             (instructionCid, instruction) <- fetchByKey @Instruction (requestors, instructionId)
             s <- case instruction.allocation of
               PLEDGE senderTransferableCid -> do
-                let holdingCid : ContractId Base.I = coerceContractId senderTransferableCid
-                holding <- fetch holdingCid
+                holding <- fetch $ toInterfaceContractId @Base.I senderTransferableCid
                 fetchAccount holding
-                pure [(getAccount holding, instructionId)]
-              _ -> pure []
+                pure . Some $ (getAccount holding, instructionId)
+              _ -> pure None
             let
               r = case instruction.approval of
-                TAKE_DELIVERY receiverAccount -> [(receiverAccount, instructionId)]
-                _ -> []
-            pure $ s ++ r
+                TAKE_DELIVERY receiverAccount -> Some (receiverAccount, instructionId)
+                _ -> None
+            pure $ catOptionals [s, r]
         -- settle
         let
           settleInstruction instructionId = do
             -- NOTE: ideally we should exerciseByKey (but currently doesn't work for interface choices).
             -- (as a workaround we could proxy the choice in the same contract)
             (instructionCid, instruction) <- fetchByKey @Instruction (requestors, instructionId)
-            exercise (toInterfaceContractId @Instruction.I instructionCid) Instruction.Execute
+            allocatedTransferable <- case instruction.allocation of
+              PLEDGE cid -> Some <$> fetch cid
+              _ -> pure None
+            transferableCidOpt <- exercise (toInterfaceContractId @Instruction.I instructionCid) Instruction.Execute
+            settledTransferable <- T.mapA fetch transferableCidOpt
+            -- undisclose transferable
+            transferableCidOpt <- fmap coerceContractId . join <$> T.mapA (\tCid -> exercise (toInterfaceContractId @Disclosure.I tCid) Disclosure.RemoveObservers with observersToRemove = (show instructionId, singleton settler)) transferableCidOpt
```

Comment:
I think we could get away with using `>>=` which means we maybe able to drop the `join`. Depends if `transferableCidOpt` is a list of optionals or not.

I'll take a quick look and see if it can be simplified or not as part of my PR once these changes are merged :)